File: parse_urdf.cpp
#include "parse_urdf.hpp"

std::vector<std::vector<double>> parse_urdf(const std::string &urdfPath) {
  std::vector<std::vector<double>> urdfParam(6, std::vector<double>(6, 0.0));
  int index = 0;
  try {
    std::ifstream file(urdfPath);
    std::vector<char> buffer((std::istreambuf_iterator<char>(file)),
                             std::istreambuf_iterator<char>());
    buffer.push_back('\0');

    rapidxml::xml_document<> doc;
    doc.parse<0>(&buffer[0]);

    rapidxml::xml_node<> *urdfRoot = doc.first_node();
    if (!urdfRoot) {
      std::cerr << "ERROR: Could not parse urdf file." << std::endl;
      return urdfParam;
    }

    for (rapidxml::xml_node<> *child = urdfRoot->first_node(); child;
         child = child->next_sibling()) {
      if (std::string(child->name()) == "joint") {
        std::string jointName = child->first_attribute("name")->value();
        index = std::stoi(std::string(1, jointName.back())) - 1;

        for (rapidxml::xml_node<> *item = child->first_node(); item;
             item = item->next_sibling()) {
          if (std::string(item->name()) == "origin") {
            std::string xyzStr = item->first_attribute("xyz")->value();
            std::string rpyStr = item->first_attribute("rpy")->value();

            // Parse xyz values
            size_t pos = 0;
            size_t endPos = 0;
            for (int i = 0; i < 3; ++i) {
              endPos = xyzStr.find(' ', pos);
              urdfParam[index][i] = std::stod(xyzStr.substr(pos, endPos - pos));
              pos = endPos + 1;
            }

            // Parse rpy values
            pos = 0;
            for (int i = 3; i < 6; ++i) {
              endPos = rpyStr.find(' ', pos);
              urdfParam[index][i] = std::stod(rpyStr.substr(pos, endPos - pos));
              pos = endPos + 1;
            }
          }
        }

      }
    }
  } catch (const rapidxml::parse_error &e) {
    std::cerr << "ERROR: Could not parse urdf file. " << e.what() << std::endl;
  }
  return urdfParam;
}

File: python_version/parse_urdf.py
import sys
import numpy as np
import xml.etree.ElementTree as ET


def parse_urdf(urdf_path):
    try:
        tree = ET.parse(urdf_path)
    except ET.ParseError:
        print('ERROR: Could not parse urdf file.')

    urdf_root = tree.getroot()
    urdf_joints = {}

    # Parse all joints and add to tree
    for child in urdf_root:
        if child.tag == 'joint':
            joint_name = child.get('name')
            for item in child:
                if item.tag == 'origin':
                    xyz = np.array(item.get('xyz').split(), dtype=float)
                    rpy = np.array(item.get('rpy').split(), dtype=float)
            urdf_joints[joint_name] = {'xyz': xyz, 'rpy': rpy}

    return urdf_joints


if __name__ == '__main__':
    if len(sys.argv) > 1:
        urdf_path = sys.argv[1]
    else:
        urdf_path = input("Please enter the URDF path: ")

    print("URDF Path:", urdf_path)

    urdf_param = parse_urdf(urdf_path)

    joint_index = 1
    for row in urdf_param:
        print(f"\nJoint Name: joint{joint_index}")
        print(f"  XYZ: {row[0]}, {row[1]}, {row[2]}")
        print(f"  RPY: {row[3]}, {row[4]}, {row[5]}")
        print()
        joint_index += 1
